ddsmt.py

```diff
@@ -262,48 +262,44 @@ def _substitute (subst_fun, substlist, superset, randomized,
     assert (g_smtformula)
     assert (substlist in (g_smtformula.subst_scopes, g_smtformula.subst_cmds,
                           g_smtformula.subst_nodes))
+    min_gran = 0.1 * sqrt(len(superset))
     nsubst_total = 0
-    gran = len(superset)
+    s = deque(superset) 
+    gran = (len(s) + 1) // 2
     
-    while gran > 0:
-        start_time = time.time()
-        if randomized:
-            subsets = [random.sample(superset, gran) for s in range(
-                       0, len(superset), gran)]
-        else:
-            subsets = [superset[s:s+gran] for s in range (
-                       0, len(superset), gran)]
-        tests_performed = 0
-        for subset in subsets:
-            if g_args.roundtime:
-                if time.time() - start_time > g_args.roundtime:
-                    _log (2, "[!!] test round timeout: reducing granularity")
-                    break
-            tests_performed += 1
+    while gran > min_gran:
+        for i in range ((len(s) + gran - 1) // gran):
+            if randomized:
+                random.shuffle(s)
             nsubst = 0
             cpy_substs = substlist.substs.copy()
             cpy_declfun_cmds = g_smtformula.scopes.declfun_cmds.copy()
-            for item in subset:
+            subset = []
+            if len(s) == 0:
+                return nsubst_total
+            for j in range (gran):
+                if len(s) == 0:
+                    break
+                item = s.popleft()
                 if not item.is_subst():
                     item.subst (subst_fun(item))
+                    subset.append(item)
                     nsubst += 1
             if nsubst == 0:
                 continue
-
+                
             _dump (g_tmpfile)
             start = time.time()
             if _test():
                 g_current_runtime = time.time() - start
-                _dump (g_args.outfile)
                 nsubst_total += nsubst
                 _log (2, "    granularity: {}, subset {} of {}, " \
-                         "substituted: {}".format(gran, tests_performed, 
-                      len(subsets), nsubst), True)
-                superset = list(set(superset) - set(subset))
+                         "substituted: {}".format(gran, i, 
+                     (len(superset)+gran-1)//gran, nsubst), True)
             else:
-                _log (2, "    granularity: {}, subset {} of {}, "\
-                         "substituted: 0".format(gran, tests_performed, 
-                      len(subsets)), True)
+                _log (2, "    granularity: {}, subset {} of {}:, " \
+                         "substituted: 0".format(gran, i, 
+                     (len(superset)+gran-1)//gran), True)
```

Comment:
Both `_log` statements print the same except that for the second case `nsubst` is always 0. I know the code was already there, but both `_log` should be merged and printed after the if/else block.